### DISCUSSION TITLE
Allow for relationship attributes to be populated from dict/JSON along with the SQLA Model itself.

### DISCUSSION THREAD

**Discussion by ccrvlh** [sqlalchemy/sqlalchemy#7202] [Usage Questions] [answered]
### Describe the use case

I currently have a complex JSON to translates to a few relationships (of all kinds, one-to-one, one-to-many and many-to-many).

I can declare the `relationship` in the SQLAlchemy (ORM) model, and use `backref`. I could normally instanciate both classes, and pass the child instance as an attribute of the parent. But as of now, I can't use kwargs to populate both the parent and children automatically. 

I confess I'm not sure that's even possible neither if it makes sense for the community, but I imagined that maybe it could be feasible to check the argument type on the `__init__` of the model. `if` is a `dict`, spread the params, something like that maybe?

### Databases / Backends / Drivers targeted

I normally use Postgres and Psycopg2.

### Example Use

```python
class Order(Base):
    __tablename__ = 'orders'

    id = db.Column(db.Integer, primary_key=True)
    order_code = db.Column(db.String)
    customer_id = db.Column(db.Integer, db.ForeignKey('customers.id'))

    customer = relationship('Customer', backref=backref('order'), uselist=False)


class Customer(Base):
    __tablename__ = 'customers'

    id = db.Column(db.Integer, primary_key=True)
    fullname = db.Column(db.String)

customer = Customer(fullname="Customer FullName")
order_model = Orders(order_code="ABC123", customer=customer)
```

This works, it would be great to make it possible to use the spread operator for child instances as well, like:
```python
# Data source (JSON/dict)
order = {
  "order_code": "Abc123",
  "customer": {
    "fullname": "Customer Fullname"
  }
)

# and then:
order_model = Orders(**order).
 # >>> this doesn't work, since as of now SQLAlchemy won't recognize a dict as initializing an instance of o child model.
```

### Additional context

Also, would be open to contribute and try to help anyway I can, would probably need some guidance though.

**Comment by zzzeek** [comment, marked as answer]
heya -

I've moved this to "usage questions" because while this is a reasonable use case, it's something that is out of scope as a first class feature in SQLAlchemy.   The overall topic here is "deserialization" and in real world use cases, usually there is more of a translational layer going on from the given dict/JSON to a model, and that's why people normally use real libraries like [marshmallow](https://github.com/marshmallow-code/marshmallow-sqlalchemy).

however, as stated, you've asked for the completely "raw" deserialize where your dictionary has a 1-1 mapping to the class constructors.   So at least for that, I can provide a short proof of concept that could be a SQLAlchemy example suite.   Example suites can be easily modified to suit specific use cases, whereas to produce a generalized extensible API is a much bigger job hence tools like marshmallow.

for a directed deserialize we can consult the [inspect interface](https://docs.sqlalchemy.org/en/14/core/inspection.html?highlight=inspect#available-inspection-targets) and use [mapper](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html#sqlalchemy.orm.Mapper) to know when a key in the JSON field refers to a relationship.  Here is a quick version of that which seems to do the general thing:

```py
def deserialize(dict_, cls):

    mp = inspect(cls)
    dict_ = dict(dict_)  # copy the dictionary to not affect the input

    for rel in mp.relationships:
        if rel.key in dict_:
            dict_[rel.key] = deserialize(dict_[rel.key], rel.mapper)

    return mp.class_(**dict_)
```

a full example follows:

```py
from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import backref
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship

from sqlalchemy import inspect


Base = declarative_base()


class Order(Base):
    __tablename__ = "orders"

    id = Column(Integer, primary_key=True)
    order_code = Column(String)
    customer_id = Column(Integer, ForeignKey("customers.id"))

    customer = relationship(
        "Customer", backref=backref("order"), uselist=False
    )

    def __repr__(self):
        # note these reprs() actually become recursive, I think the Python
        # interpreter has a special feature to prevent repr() from going into
        # a recursive loop so it works out
        return "Order(%s)" % (
            ", ".join(
                (
                    "%s=%r" % (attr.key, getattr(self, attr.key))
                    for attr in inspect(self).attrs
                )
            )
        )


class Customer(Base):
    __tablename__ = "customers"

    id = Column(Integer, primary_key=True)
    fullname = Column(String)

    def __repr__(self):
        return "Customer(%s)" % (
            ", ".join(
                (
                    "%s=%r" % (attr.key, getattr(self, attr.key))
                    for attr in inspect(self).attrs
                )
            )
        )


def deserialize(dict_, cls):

    mp = inspect(cls)
    dict_ = dict(dict_)  # copy the dictionary to not affect the input

    for rel in mp.relationships:
        if rel.key in dict_:
            dict_[rel.key] = deserialize(dict_[rel.key], rel.mapper)

    return mp.class_(**dict_)


order_dict = {
    "order_code": "Abc123",
    "customer": {"fullname": "Customer Fullname"},
}

order = deserialize(order_dict, Order)

print(order)

# customer = Customer(fullname="Customer FullName")
# order_model = Orders(order_code="ABC123", customer=customer)
```

then we get the repr output:

```
$ python test3.py 
Order(customer=Customer(id=None, fullname='Customer Fullname', order=[Order(customer=Customer(id=None, fullname='Customer Fullname', order=[...]), id=None, order_code='Abc123', customer_id=None)]), id=None, order_code='Abc123', customer_id=None)
```

**Reply by CaselIT** [reply]
An option may also be to pass a custom `__init__` to `declarative_base()`, that does something similar to `deserialize`

**Reply by ccrvlh** [reply]
@zzzeek thanks a lot for that, it makes sense, I'll try to implement it, probably using @CaselIT  suggestion.
I'm actually using Pydantic, but it does not have the integration directly (normally it just uses the `.dict()` method to pass arguments to SQLAlchemy Models as `kwargs`. I've never used marshamallow, that may also be an option. Appreciate the example, thanks.

**Reply by CaselIT** [reply]
It seems that pydantic has also some support for this: https://pydantic-docs.helpmanual.io/usage/models/#orm-mode-aka-arbitrary-class-instances

never used it though, so I'm not sure if it fits your needs

**Reply by ccrvlh** [reply]
This works the other way round: from SQLAlchemy to create Pydantic models. But normally instanciating SQLAlchemy models is model through `kwargs` with `MyModel(**PydanticSchema.dict())` method. I might be missing something though, maybe there's a another way that I don't know of. Will ask the pydantic community as well.

**Reply by CaselIT** [reply]
Oh ok. I though it also did the inverse
